### PR TITLE
docs: Add builtins.md

### PR DESCRIPTION
**Pull request by juliaogris**
Add builtins.md with docs for I/O builtin functions. In follow up
commits we will extend builtins.md with the following sections:

- Types
- Map
- Program Control
- Conversion
- String
- Random
- Math
- Error
- UI
- Event Handlers

Add a `doctest` target to the Makefile run with `make ci` checking for
non-dirty git status of the relevant markdown files.

Update github CI environment with the environment variable `TERM=vt100`.
Without this environment variable set `cls` fails and without that we
cannot run doctest on CI.

---

Rendered markdown:
https://github.com/foxygoat/evy/blob/builtin-docs/docs/builtins.md

Current WIP:
https://github.com/foxygoat/evy/blob/builtin-types/docs/builtins.md

### CODE REVIEW

File: docs/builtins.md [hunk 1]
```diff
@@ -0,0 +1,718 @@
...
+`print` prints its arguments to the output area, each separated by a
+single space and terminated by a newline character. If no arguments are
+provided it only prints the newline character. 
+
+A tab is printed for the `\t` control character. A newline is printed
+for the `\n` control character. Quotes in string literals are escaped
+with backslash `\`, for example `print "Here's a tab: \t.\nShe said: \"Thank you!\""`.
+
+In a browser environment `print` typically prints to a DOM Element wired
+to the evy runtime. When using evy as Command Line Interface, `print`
```
Comment (camh-) [inline docs/builtins.md:47]: "When using evy as Command Line Interface" doesn't read very well. Perhaps "When running evy from the command line interface"

File: docs/builtins.md [hunk 2]
```diff
@@ -0,0 +1,718 @@
...
+prints to Stdout.
```
Comment (camh-) [inline docs/builtins.md:48]: s/Stdout/standard out/

File: docs/builtins.md [hunk 3]
```diff
@@ -0,0 +1,718 @@
...
+
+    read:string
+
+`read` returns a string that contains the line of input that the user
+entered up until, excluding the newline character. It is a blocking
+functions, which means that it will not return until the user has
+entered a line of input and pressed the Enter key.
+
+In a browser environment `read` typically reads from a DOM Textarea
+Element wired to the evy runtime. When using evy as Command Line
```
Comment (camh-) [inline docs/builtins.md:81]: same here: When running evy from the command line interface...

File: docs/builtins.md [hunk 4]
```diff
@@ -0,0 +1,718 @@
...
+Interface, `read` reads from Stdin.
```
Comment (camh-) [inline docs/builtins.md:82]: s/Stdin/standard in/

File: docs/builtins.md [hunk 5]
```diff
@@ -0,0 +1,718 @@
...
+```
+Bye
+```
+
+#### Reference 
+
+    cls
+
+In a browser environment `cls` typically clears the DOM Element wired to
+the evy runtime which is used for printing. When using evy as Command
```
Comment (camh-) [inline docs/builtins.md]: again: from the command line interface.

File: docs/builtins.md [hunk 6]
```diff
@@ -0,0 +1,718 @@
...
+Line Interface, `cls` clears the terminal, similar to the Unix `clear`
+or Windows `cls` commands.
+
+### `printf`
+
+`printf` stands for print formatted.
+
+`printf` prints its arguments to the output area according to a _format_
+string. The format string is the first argument, and it
+contains _specifiers_. Specifiers are placeholder that start with a
```
Comment (camh-) [inline docs/builtins.md:117]: I think you could remove "are placeholders that", so it reads: Specifiers start with a ...

Introducing the word "placeholder" when not used elsewhere doesn't really add anything here I think.

File: docs/builtins.md [hunk 7]
```diff
@@ -0,0 +1,718 @@
...
+| Specifier | Description |
+| --------- | ------------- |
+| `%v`      | the argument in its default format  |
+| `%q`      | a double-quoted string |
+| `%%`      | a percent sign `%`  |
+
+#### Example
+
+```evy
+printf "%%v string: %v\n" "abc"
```
Comment (camh-) [inline docs/builtins.md:133]: I think using `%%` in each example is perhaps a little confusing. I get what you are going for with the output, but I wonder if there's a better way to do these examples?

File: docs/builtins.md [hunk 8]
```diff
@@ -0,0 +1,718 @@
...
+%v array: [1 2 3]
+%v map: {a:1 b:2}
+```
+
+#### Reference 
+
+    printf format:string a:any...
+
+`printf` prints its arguments to the output area according to
+the _format_ string that is represented by the first argument.
```
Comment (camh-) [inline docs/builtins.md]: "that is represented by the first argument is a bit wordy". (I think "represented by" is a longer way of just saying "is". Perhaps

`printf` prints its first argument (the format string) to the output area, replacing
any specifiers in the format string with the following arguments. A _specifier_ starts
with a `%` and specifies how the value of one of the remaining arguments is
formatted. Each specifier (apart from the `%%` specifier, see below) consumes the
next unconsumed argument in order. For example, ...

**Review by camh- — COMMENTED** [review]
Since you seem to be iterating on this as I review, here are some comments so far,...

File: docs/builtins.md [hunk 9]
```diff
@@ -0,0 +1,718 @@
...
+`print` prints the arguments given to it to the output area. It separates them by
+a single space and outputs a newline character at the end.
+
+#### Example 
+
+```evy
+print "Hello"
+print 2 true "blue"
+print "array:" [1 2 3]
+print "map:" {name: "Scholl" age: 21}
```
[the remark this answers is not in the document]
Comment (juliaogris) [reply docs/builtins.md:24]: nice, done, thanks.

File: docs/builtins.md (hunk 1 above)
Comment (juliaogris) [reply docs/builtins.md:47]: thanks, agreed!

File: docs/builtins.md (hunk 2 above)
Comment (juliaogris) [reply docs/builtins.md:48]: Done.

File: docs/builtins.md [hunk 10]
```diff
@@ -0,0 +1,730 @@
...
+| Verb    | Description |
+| ------- | ------------- |
+| %f      | default width, default precision |
+| %7f     | width 7, default precision |
+| %.2f    | default width, precision 2 |
+| %7.2f   | width 7, precision 2 |
+| %7.f    | width 7, precision 0 |
+
+
+The `-` flag tells `printf` to pad the output with spaces on the right,
```
Comment (camh-) [inline docs/builtins.md:205]: I think "flag" is the wrong word to use here. I would perhaps say:

If the width/precision is preceded by a `-`, the value is padded with spaces on the
right rather than the left. If it is preceded by a `0`, the value is padded with leading
zeros rather than spaces.

File: docs/builtins.md [hunk 11]
```diff
@@ -0,0 +1,730 @@
...
+rather than the left. The `0` flag tells `printf` to pad the output
+with leading zeros, rather than spaces.
+
+The width number and the align-left flag  `-` can be used with all valid
```
Comment (camh-) [inline docs/builtins.md:209]: nit: double space after "flag"

File: docs/builtins.md (hunk 3 above)
Comment (juliaogris) [reply docs/builtins.md:81]: done.

File: docs/builtins.md (hunk 4 above)
Comment (juliaogris) [reply docs/builtins.md:82]: done.

File: docs/builtins.md (hunk 5 above)
Comment (juliaogris) [reply docs/builtins.md]: done.

File: docs/builtins.md (hunk 6 above)
Comment (juliaogris) [reply docs/builtins.md:117]: ok, agreed, thank you!

File: docs/builtins.md (hunk 8 above)
Comment (juliaogris) [reply docs/builtins.md]: ok, agreed.

File: docs/builtins.md (hunk 11 above)
Comment (camh-) [inline docs/builtins.md:209]: To get rid of "flag" here, I'd say something like:

A specifier width and alignment prefix (`-` or `0`) can be used with all valid specifiers. For example:

And then the examples should have a non %f example I think, since that "for example" seems to be about that.

File: docs/builtins.md [hunk 12]
```diff
@@ -0,0 +1,730 @@
...
+specifiers. For example:
+
+```evy
+printf "right:  |%7.2f|\n" 1
+printf "left:   |%-7.2f|\n" 1
+printf "zeropad:|%07.2f|\n" 1
+```
+Output
+```
+1 2.000000 3
```
Comment (camh-) [inline docs/builtins.md:219]: this line of output seems to be left over from some earlier example

File: docs/builtins.md [hunk 13]
```diff
@@ -0,0 +1,730 @@
...
+
+`typeof` returns the type of the argument as string value.
+
+#### Example
+
+```evy
+a:any
+a = "abcd"
+t := typeof a
+print "typof \"abcd\":" t
```
Comment (camh-) [inline docs/builtins.md:273]: s/typof/typeof/ - and in the output section below too.

File: docs/builtins.md [hunk 14]
```diff
@@ -0,0 +1,730 @@
...
+
+t = typeof {kind:true strong:true}
+print "typeof {a:3 b:4 c:5}:" t
```
Comment (camh-) [inline docs/builtins.md:276]: it looks like `{a:3 b:4 c:5}` is wrong here and should be `{kind:true strong: true}`

File: docs/builtins.md [hunk 15]
```diff
@@ -0,0 +1,730 @@
...
+
+t = typeof [[1 2] [3 4]]
+print "typeof [1 2 true]:" t
```
Comment (camh-) [inline docs/builtins.md:279]: same here - the value in the string is not what `typeof` is being run on

File: docs/builtins.md [hunk 16]
```diff
@@ -0,0 +1,730 @@
...
+exit function is called and print the status code to the output area.
+This makes it easy to identify the source of any errors that occur in
+your program.
+
+## Conversion
+
+### `str2num`
+
+`str2num` converts a string to a number. If the string is not a valid
+number, it returns `0` and sets the global `err` to true.
```
Comment (camh-) [inline docs/builtins.md:435]: Hmmm, this is the first mention of  "global `err`". Perhaps that needs an introduction, or a section that you can link to at least?

Perhaps you meant "sets the global variable `err`" which is probably enough to clear this up.

File: docs/builtins.md [hunk 17]
```diff
@@ -0,0 +1,730 @@
...
+#### Reference 
+
+    str2num:num s:string
+
+TBD
+
+### `str2bool`
+
+`str2bool` converts a string to a boolean. If the string is not a valid
+boolean, it returns `0` and sets the global `err` to true.
```
Comment (camh-) [inline docs/builtins.md:462]: same here: global variable

File: docs/builtins.md [hunk 18]
```diff
@@ -0,0 +1,730 @@
...
+
+TBD
+
+### `sprintf`
+
+`sprintf` stands for print formatted to string. 
+
+`sprintf` returns a string representation of its arguments according to
+a _format_ string. Formatting in `sprintf` and `printf` work the exact
+same way, `printf` for further details.
```
Comment (camh-) [inline docs/builtins.md:523]: see `printf`

File: docs/builtins.md (hunk 7 above)
Comment (juliaogris) [reply docs/builtins.md:133]: ok, pondering.

File: docs/builtins.md (hunk 12 above)
Comment (juliaogris) [reply docs/builtins.md:219]: done

File: docs/builtins.md (hunk 13 above)
Comment (juliaogris) [reply docs/builtins.md:273]: done.

File: docs/builtins.md (hunk 14 above)
Comment (juliaogris) [reply docs/builtins.md:276]: done.

File: docs/builtins.md (hunk 11 above)
Comment (juliaogris) [reply docs/builtins.md:209]: ok, great! thank you. all fixed up (I believe).

File: docs/builtins.md (hunk 10 above)
Comment (juliaogris) [reply docs/builtins.md:205]: done.

File: docs/builtins.md (hunk 11 above)
Comment (juliaogris) [reply docs/builtins.md:209]: flag removed.

File: docs/builtins.md (hunk 15 above)
Comment (juliaogris) [reply docs/builtins.md:279]: oops, sorry, fixed.

File: docs/builtins.md (hunk 17 above)
Comment (juliaogris) [reply docs/builtins.md:462]: pre-declared, global variable ?
fixed with what you suggested.

File: docs/builtins.md (hunk 16 above)
Comment (juliaogris) [reply docs/builtins.md:435]: I think a section on err would be a good idea actually.
Maybe with program control?

File: docs/builtins.md (hunk 18 above)
Comment (juliaogris) [reply docs/builtins.md:523]: yes!

File: docs/builtins.md [hunk 19]
```diff
@@ -0,0 +1,806 @@
...
+printf "There will be unicorns eating lollipops: %v.\n\n" unicorns
+
+printf "They said: %q\n" "wow"
+printf "Array: %v\n" [1 2 3]
+printf "Map: %v\n" {a:1 b:2}
+```
+
+Output
+```
+%v string: abc
```
Comment (camh-) [inline docs/builtins.md:149]: This output needs to be updated for the examples above.

File: docs/builtins.md [hunk 20]
```diff
@@ -0,0 +1,806 @@
...
+has {a:1} "X": false
+```
+
+#### Reference 
+
+    has:bool map:{} key:string
+
+The `has` function takes two arguments: a map and a key. The function
+returns true if the map has the key, and false if the map does not have
+the key. The map can have any value type, such as `{}num` or `{}
```
Comment (camh-) [inline docs/builtins.md:336]: nit: I would not split the type across lines. it should still fit and be shorter than the previous line if joined.

File: docs/builtins.md [hunk 21]
```diff
@@ -0,0 +1,806 @@
...
+exit function is called and print the status code to the output area.
+This makes it easy to identify the source of any errors that occur in
+your program.
+
+## Conversion
+
+### `str2num`
+
+`str2num` converts a string to a number. If the string is not a valid
+number, it returns `0` and sets the global variable `err` to `true`.
```
Comment (camh-) [inline docs/builtins.md:437]: In the other places, you've use "global `err` variable". Should probably do the same here.

File: docs/builtins.md (hunk 16 above)
Comment (camh-) [reply docs/builtins.md]: Perhaps. There isn't much to actually say - "There is a pre-declared global variable `err` that gets set by some functions". Perhaps it is just enough to refer to the "global `err` variable" as you have now.

One thing that is missing is what happens to the variable when there is no error in the functions that set it. Is it the programmers responsibility to reset the var to `false`?

Comment (juliaogris) [reply docs/builtins.md]: well i guess there is also `errmsg`, e.g.:
```
n := str2num "asdf"
print n err errmsg
```
outputs:
```
0 true str2num: cannot parse asdf
```
so maybe it deserves a separate section?

File: docs/builtins.md (hunk 19 above)
Comment (juliaogris) [reply docs/builtins.md]: oops, sorry! done.

File: docs/builtins.md (hunk 20 above)
Comment (juliaogris) [reply docs/builtins.md:336]: done.

File: docs/builtins.md (hunk 21 above)
Comment (juliaogris) [reply docs/builtins.md:437]: fixed, thanks.

File: docs/builtins.md [hunk 22]
```diff
@@ -0,0 +1,225 @@
...
+
+#### Reference 
+
+    print a:any...
+
+The `print` function prints its arguments to the output area, each
+separated by a single space and terminated by a newline character. If
+no arguments are provided it only prints the newline character. 
+
+A tab is printed for the `\t` control character. A newline is printed
```
Comment (camh-) [inline docs/builtins.md:42]: technically the tab is the "control character". `\t` is an escape sequence or backslash sequence or something like that

File: docs/builtins.md [hunk 23]
```diff
@@ -0,0 +1,225 @@
...
+for the `\n` control character. Quotes in string literals are escaped
+with backslash `\`, for example `print "Here's a tab: \t.\nShe said: \"Thank you!\""`.
+
+In a browser environment `print` typically prints to a DOM Element wired
```
Comment (camh-) [inline docs/builtins.md:46]: "typically prints to a DOM Element wired to the evy runtime". Why is this "typically"? Could you just say "outputs to the output area" or something? I know the code could be changed to do other things, but this is what it does right now.

File: docs/builtins.md [hunk 24]
```diff
@@ -0,0 +1,225 @@
...
+#### Reference 
+
+    read:string
+
+The `read` function returns a string that contains the line of input
+that the user entered up until, excluding the newline character. It is
+a blocking functions, which means that it will not return until the
+user has entered a line of input and pressed the Enter key.
+
+In a browser environment `read` typically reads from a DOM Textarea
```
Comment (camh-) [inline docs/builtins.md:80]: same comment about "typically..."

File: docs/builtins.md [hunk 25]
```diff
@@ -0,0 +1,225 @@
...
+by the remaining arguments in the given order. For example, the
+following code `printf "first: %s, second: %s" "A" "B"` prints `first:
+A, second: B`.
+
+Full list of valid specifiers in Evy:
+
+| Specifier | Description |
+| --------- | ------------- |
+| %v        | the argument in a default format  |
+| %%        | a literal percent sign `%`; consumes no value  |
```
Comment (camh-) [inline docs/builtins.md:176]: I would put this one first or last as it is different to the others.

**Review by camh- — APPROVED** [review]
🫛 Few minor comments, but all good

File: docs/builtins.md (hunk 22 above)
Comment (juliaogris) [reply docs/builtins.md:42]: ok. I have reworded the whole thing a little, hope that's ok.

File: docs/builtins.md (hunk 23 above)
Comment (juliaogris) [reply docs/builtins.md:46]: ok, done, thank you.

File: docs/builtins.md (hunk 24 above)
Comment (juliaogris) [reply docs/builtins.md:80]: ok, done.

File: docs/builtins.md (hunk 25 above)
Comment (juliaogris) [reply docs/builtins.md:176]: fair, i put it at the end like above. seems also more suited there to me.

**Review by juliaogris — COMMENTED** [review]
@camh- so i think i have addressed all comments, but could have been a 🐾 and drink to many....
i think the doctest script fails samples in the reference section, not sure, haven't dug...
